nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,142 @@ with lib;
 
 let
   cfg = config.networking.wireless.iwd;
+
+  encodeFileName = pkgs.writeScript "encode" ''
+    #! ${pkgs.runtimeShell} -e
+    echo "=$(printf "$1" | ${pkgs.unixtools.xxd}/bin/xxd -pu)"
+  '';
+
+  # If a network consists of alphanumeric chars and `_` and `-`, the network will be stored
+  # as `/var/lib/iwd/<ssid>.<ext>`. In any other case a hex-encoded version of the SSID will
+  # be stored with an `=` as prefix.
+  encodeIfNeeded = file: ext: let
+    allowedChars = stringToCharacters "01234556789abcdefghijklmnopqrstuvwxyz-_ ";
+    skipEncoding = all (char: elem (toLower char) allowedChars) (stringToCharacters file);
+  in
+    "/var/lib/iwd/${if skipEncoding then file else "$(${encodeFileName} \"${file}\")"}.${ext}";
+
+  # Basically `types.str`, but the `check` function validates the string-length as well.
+  fixedStringLength = min: max: mkOptionType {
+    name = "str-with-length";
+    description = if min == max
+      then "string with exactly ${toString min} chars."
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
+  };
+
+  # The file extension of the network-config determines the type of the network. Currently,
+  # we support public networks (`open`) and networks protected with a passphrase (`psk`).
+  # Anything else, like 802.1x networks are not supported by this module.
+  determineType = cfg: if cfg.psk != null || cfg.passphrase != null then "psk" else "open";
+
+  # Generates a name-value-pair with configurations for a SSID. For further information
+  # about the configuration options, please refer to the upstream docs:
+  #
+  # https://iwd.wiki.kernel.org/networkconfigurationsettings
+  predefinedNetworks = flip mapAttrs cfg.networks (name: cfg: {
+    extension = determineType cfg;
+
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" (generators.toINI {} ({
+      Settings.AutoConnect = cfg.autoConnect;
+      Settings.Hidden = cfg.hidden;
+    } // (optionalAttrs (cfg.psk != null) {
+      Security.PreSharedKey = cfg.psk;
+    }) // (optionalAttrs (cfg.passphrase != null) {
+      Security.Passprase = cfg.passphrase;
+    })));
+  });
 in {
-  options.networking.wireless.iwd.enable = mkEnableOption "iwd";
+  options.networking.wireless.iwd = {
+    enable = mkEnableOption "iwd";
+
+    interfaces = mkOption {
+      type = types.nullOr (types.listOf types.str);
+      example = [ "wlp2s0" ];
+      apply = x: if x == null then x else concatStringsSep " " x;
+      default = null;
```

Comment:
Generally lists should be avoided as types, because they compose very badly. E.g. once you have a list of say 10 interfaces, if you want to remove one, you have to `mkForce` the 9 of them that you want which is pretty bad. Instead I suggest using `types.attrsOf types.bool`, allowing you to use `{ wlp2s0 = true; }` to enable one and `{ wlp2s0 = mkForce false; }` to disable it anywhere else again.

In addition, avoid using `apply`, since it makes reading config values for other modules tricky. E.g. in this case, to figure out whether any interface is enabled, they'd have to split `config.networking.wireless.iwd` on empty chars first.

Reply:
> Generally lists should be avoided as types, because they compose very badly. E.g. once you have a list of say 10 interfaces, if you want to remove one, you have to mkForce the 9 of them that you want, which is rather ugly

In theory I agree with your reasoning, but I don't think we should do that here: I mean, how many wireless network interfaces does an average machine have? Also, you probably have to specify those per machine(-type) either way due to the variety of different names due to udev's predictable interface-naming feature (which we have enabled by default).

I think that for such a small value, the (IMHO more intuitive) `listOf`type is preferrable.

Reply:
As this isn't something I forsee anyone overriding anywhere but the same place it's set, I think it's perfectly reasonable to leave it as a list, though I also agree generally with the idea

Reply:
Imo making this an `attrsOf bool` instead isn't a big cost, and it would set a good sample for other modules.

But also, a problem I haven't mentioned yet: `listOf` arguably breaks determinism of the module system, because it makes the order of elements depend on the order of module imports: Changing `imports = [ ./foo.nix ./bar.nix ];` doesn't have to evaluate to the same thing as `imports = [ ./bar.nix ./foo.nix ];` for `listOf` options. I previously had to fix a bug due to this with https://github.com/NixOS/nixpkgs/pull/77088.

I think `listOf` should really only be used when it's a single list value that's assigned, e.g. `services.xserver.synaptics.buttonsMap`. And I actually plan to make an RFC to slowly deprecate `listOf` because of the problems I mentioned above.

Reply:
> Imo making this an attrsOf bool instead isn't a big cost, and it would set a good sample for other modules.

I don't know tbh: I personally think that this interface is a little bit too clumsy.

> But also, a problem I haven't mentioned yet: listOf arguably breaks determinism of the module system

OTOH you sometimes have to depend on the order of values in a list which doesn't work anymore with attrs as those are always ordered alphabetically, right? Anyway, I don't see why this issue applies here.

Reply:
Usually the order doesn't matter yeah. But it still messes with determinism a bit, since flipping imports can lead to different final system derivations.

Anyways, I have a plan for fixing these problems, let's leave this at `listOf str` for now then :)